File: reference/chunking.py
import torch
from util.save_reference import save_reference
from typing import Tuple

def chunking(tensor: torch.Tensor, chunk_height: int, chunk_width: int, pad: int = 0) -> torch.Tensor:
    """
    Reshapes a 4D tensor (batch, height, width, channels) into chunks.

    Args:
        tensor (torch.Tensor): Input tensor of shape (batch, height, width, channels).
        chunk_height (int): The height of each chunk.
        chunk_width (int): The width of each chunk.
        pad (int): Value used for padding if the dimensions are not perfectly divisible
                   by chunk_height/chunk_width. Defaults to 0.

    Returns:
        torch.Tensor: A 3D tensor of shape (batch, num_total_chunks, chunk_height * chunk_width).
                      Chunks are ordered by width_step, then height_step, then channel.
    """
    batch, height, width, channels = tensor.shape

    if chunk_height <= 0 or chunk_width <= 0:
        raise ValueError("chunk_height and chunk_width must be positive.")

    num_h_steps = (height + chunk_height - 1) // chunk_height
    height_padded = num_h_steps * chunk_height
    num_w_steps = (width + chunk_width - 1) // chunk_width
    width_padded = num_w_steps * chunk_width

    height_needed_pad = height_padded - height
    width_needed_pad = width_padded - width

    # Padding format for torch.nn.functional.pad is (pad_left_dimN, pad_right_dimN, ...).
    # For a 4D tensor (B, H, W, C), tensor.pad expects padding for last 3 dims (C, W, H)
    # So, padding tuple is (pad_C_left, pad_C_right, pad_W_left, pad_W_right, pad_H_left, pad_H_right)
    padding = (0, 0,  # Channels padding (none)
               0, width_needed_pad,   # Width padding (only on the right)
               0, height_needed_pad)  # Height padding (only at the bottom)

    padded_tensor = torch.nn.functional.pad(tensor, padding, mode="constant", value=pad)

    # Reshape to (batch, num_h_steps, chunk_height, num_w_steps, chunk_width, channels)
    reshaped = padded_tensor.reshape(batch, num_h_steps, chunk_height, num_w_steps, chunk_width, channels)

    # Permute to (batch, num_w_steps, num_h_steps, channels, chunk_height, chunk_width)
    # This order ensures that when we flatten, chunks are grouped by w_step, then h_step, then channel.
    # Original indices: (0:B, 1:N_h, 2:C_h, 3:N_w, 4:C_w, 5:Chan)
    # New order:        (0:B, 3:N_w, 1:N_h, 5:Chan, 2:C_h, 4:C_w)
    permuted = reshaped.permute(0, 3, 1, 5, 2, 4)

    # Reshape to (batch, num_total_chunks, chunk_area)
    # num_total_chunks = num_w_steps * num_h_steps * channels
    # chunk_area = chunk_height * chunk_width
    output_tensor = permuted.reshape(batch, num_w_steps * num_h_steps * channels, chunk_height * chunk_width)

    return output_tensor

def create_chunking(
    shape: Tuple[int, int, int, int],
    chunk_height: int,
    chunk_width: int,
    pad_value: int = 0,
    dtype: torch.dtype = torch.float32,
    dir: str = "data",
    name: str = "chunking",
):
    """
    Create a tensor, perform chunking, and save the original and chunked tensors.

    Args:
        shape (Tuple[int, int, int, int]): Shape of the tensor (batch, height, width, channels).
        chunk_height (int): Height of each chunk.
        chunk_width (int): Width of each chunk.
        pad_value (int): Value to use for padding. Default is 0.
        dtype (torch.dtype): Data type of the tensor. Default is torch.float32.
        dir (str): Directory to save the reference tensors. Default is "data".
        name (str): Base name for the reference tensor files. Default is "chunking".
    """
    if len(shape) != 4:
        raise ValueError("Input shape must be 4D (batch, height, width, channels).")
    if any(s <= 0 for s in shape):
        raise ValueError("All dimensions in shape must be positive.")
    x = torch.randn(shape, dtype=dtype)


    save_reference(x, dir, f"{name}_x")

    chunked_tensor = chunking(x, chunk_height, chunk_width, pad=pad_value)
    save_reference(chunked_tensor, dir, f"{name}_result_ch{chunk_height}_cw{chunk_width}_pad{pad_value}")

if __name__ == "__main__":
    create_chunking(
        shape=(2, 32, 32, 3),  # batch, height, width, channels
        chunk_height=8,
        chunk_width=8,
        pad_value=0,
        dtype=torch.float32,
        dir="data",
        name="chunking_32x32_ch8cw8_float"
    )

    create_chunking(
        shape=(1, 5, 7, 1),   # batch, height, width, channels
        chunk_height=2,
        chunk_width=3,
        pad_value=0,          # Pad with zero
        dtype=torch.float32,
        dir="data",
        name="chunking_5x7_ch2cw3_int_pad0"
    )

    create_chunking(
        shape=(1, 3, 3, 2),   # batch, height, width, channels
        chunk_height=2,
        chunk_width=2,
        pad_value=5,          # Pad with five
        dtype=torch.float32,
        dir="data",
        name="chunking_3x3_ch2cw2_float16_pad5"
    )

    # Example 4: Single large chunk (entire feature map)
    create_chunking(
        shape=(1, 7, 7, 1),
        chunk_height=7,
        chunk_width=7,
        pad_value=0,
        dtype=torch.float32,
        dir="data",
        name="chunking_7x7_ch7cw7_full"
    )

    # Example 5: 1x1 chunks (effectively a flatten operation with channel interleaving)
    create_chunking(
        shape=(1, 2, 2, 2),
        chunk_height=1,
        chunk_width=1,
        pad_value=0,
        dtype=torch.float32,
        dir="data",
        name="chunking_2x2_ch1cw1_int8"
    )
    print("Chunking reference tensors created in 'data' directory.")

File: src/functions/chunking.rs
use ndarray::{Array, ArrayD, Axis, Dimension, IxDyn, ShapeError, s};
use num_traits::{AsPrimitive, Num, Zero}; // Added AsPrimitive for pad_value if needed

#[derive(Debug)]
pub enum ChunkingError {
    InvalidInputShape(String),
    InvalidChunkSize(String),
    NdarrayShapeError(ShapeError),
    // Add other error types as needed
}

impl From<ShapeError> for ChunkingError {
    fn from(err: ShapeError) -> ChunkingError {
        ChunkingError::NdarrayShapeError(err)
    }
}

/// Reshapes a 4D tensor (batch, height, width, channels) into chunks.
///
/// Args:
///     tensor: Input tensor of shape (batch, height, width, channels).
///     chunk_height: The height of each chunk.
///     chunk_width: The width of each chunk.
///     pad_value: Value used for padding if the dimensions are not perfectly divisible.
///
/// Returns:
///     A 3D tensor of shape (batch, num_total_chunks, chunk_height * chunk_width).
///     Chunks are ordered by width_step, then height_step, then channel.
pub fn chunking<T>(
    tensor: &ArrayD<T>,
    chunk_height: usize,
    chunk_width: usize,
    pad_value: T,
) -> Result<ArrayD<T>, ChunkingError>
where
    T: Num + Copy + Zero, // Zero trait for pad_value, Copy for element copying
{
    // 1. Get shape and validate
    if tensor.ndim() != 4 {
        return Err(ChunkingError::InvalidInputShape(
            "Input tensor must be 4D (batch, height, width, channels)".to_string(),
        ));
    }
    let shape = tensor.shape(); // shape() returns &[usize]
    let batch = shape[0];
    let height = shape[1];
    let width = shape[2];
    let channels = shape[3];

    if chunk_height == 0 || chunk_width == 0 {
        return Err(ChunkingError::InvalidChunkSize(
            "chunk_height and chunk_width must be positive.".to_string(),
        ));
    }

    // 2. Calculate padding dimensions
    let num_h_steps = (height + chunk_height - 1) / chunk_height;
    let height_padded = num_h_steps * chunk_height;
    let num_w_steps = (width + chunk_width - 1) / chunk_width;
    let width_padded = num_w_steps * chunk_width;

    // let _height_needed_pad = height_padded - height; // For reference
    // let _width_needed_pad = width_padded - width;   // For reference

    // 3. Pad the tensor
    let padded_dim_slice = [batch, height_padded, width_padded, channels];
    let mut padded_tensor = Array::from_elem(IxDyn(&padded_dim_slice), pad_value);

    // Create a view of the original tensor with explicit 4D shape for assignment
    let original_tensor_view_4d = tensor.view().into_dimensionality::<ndarray::Ix4>()?; // Convert &ArrayD<T> to ArrayView<T, Ix4>

    // Copy original data into the padded tensor
    padded_tensor
        .slice_mut(s![.., 0..height, 0..width, ..])
        .assign(&original_tensor_view_4d);

    // 4. Reshape to (batch, num_h_steps, chunk_height, num_w_steps, chunk_width, channels)
    let reshaped_dims_slice = [
        batch,
        num_h_steps,
        chunk_height,
        num_w_steps,
        chunk_width,
        channels,
    ];
    let reshaped = padded_tensor.into_shape(IxDyn(&reshaped_dims_slice))?;

    // 5. Permute
    // Python: permuted = reshaped.permute(0, 3, 1, 5, 2, 4)
    // Original indices: (0:B, 1:N_h, 2:C_h, 3:N_w, 4:C_w, 5:Chan)
    // New order:        (0:B, 3:N_w, 1:N_h, 5:Chan, 2:C_h, 4:C_w)
    let permute_order: &[usize] = &[0, 3, 1, 5, 2, 4];
    // permute_axes returns an ArrayView. We need an owned array for the next reshape.
    let permuted = reshaped.permuted_axes(permute_order).to_owned();

    // 6. Reshape to (batch, num_total_chunks, chunk_area)
    let num_total_chunks = num_w_steps * num_h_steps * channels;
    let chunk_area = chunk_height * chunk_width;

    let output_dims_slice = [batch, num_total_chunks, chunk_area];
    let output_tensor = permuted
        .as_standard_layout()
        .to_owned()
        .into_shape(IxDyn(&output_dims_slice))?;

    Ok(output_tensor) // output_tensor is Array<T, IxDyn>, which is type alias for ArrayD<T>
}
